libraries/node-core-library/src/test/Async.test.ts

```diff
@@ -459,3 +459,141 @@ describe(Async.name, () => {
     });
   });
 });
+
+describe(AsyncQueue.name, () => {
+  it('Can enqueue and dequeue items', async () => {
+    const expectedItems: Set<number> = new Set([1, 2, 3]);
+    const expectedSeenItems: number = 3;
+
+    let seenItems: number = 0;
+    const queue: AsyncQueue<number> = new AsyncQueue<number>(expectedItems);
+
+    for await (const [item, callback] of queue) {
+      seenItems++;
+      expect(expectedItems.has(item)).toBe(true);
+      expectedItems.delete(item);
+      callback();
+    }
+
+    expect(seenItems).toEqual(expectedSeenItems);
+    expect(expectedItems.size).toEqual(0);
+  });
+
+  it('Can dynamically enqueue and dequeue items', async () => {
+    const expectedItems: Set<number> = new Set([1, 2, 3]);
+    const expectedAdditionalItems = new Set([4, 5, 6]);
+    const expectedSeenItems: number = 6;
+
+    let seenItems: number = 0;
+    const queue: AsyncQueue<number> = new AsyncQueue<number>(expectedItems);
+
+    for await (const [item, callback] of queue) {
+      seenItems++;
+      if (item < 4) {
+        expect(expectedItems.has(item)).toBe(true);
+        expectedItems.delete(item);
+        queue.push(item + 3);
+      } else {
+        expect(expectedAdditionalItems.has(item)).toBe(true);
+        expectedAdditionalItems.delete(item);
+      }
+      callback();
+    }
+
+    expect(seenItems).toEqual(expectedSeenItems);
+    expect(expectedItems.size).toEqual(0);
+    expect(expectedAdditionalItems.size).toEqual(0);
+  });
+
+  it('Can enqueue and dequeue items concurrently', async () => {
+    const expectedItems: Set<number> = new Set([1, 2, 3]);
+    const expectedSeenItems: number = 3;
+
+    let seenItems: number = 0;
+    const queue: AsyncQueue<number> = new AsyncQueue<number>(expectedItems);
+
+    await Async.forEachAsync(
+      queue,
+      async ([item, callback]) => {
+        seenItems++;
+        expect(expectedItems.has(item)).toBe(true);
+        expectedItems.delete(item);
+        callback();
+      },
+      {
+        concurrency: 10
+      }
+    );
+
+    expect(seenItems).toEqual(expectedSeenItems);
+    expect(expectedItems.size).toEqual(0);
+  });
+
+  it('Can dynamically enqueue and dequeue items concurrently', async () => {
+    const expectedItems: Set<number> = new Set([1, 2, 3]);
+    const expectedAdditionalItems = new Set([4, 5, 6]);
+    const expectedSeenItems: number = 6;
+
+    let seenItems: number = 0;
+    const queue: AsyncQueue<number> = new AsyncQueue<number>(expectedItems);
+
+    await Async.forEachAsync(
+      queue,
+      async ([item, callback]) => {
+        seenItems++;
```

Comment:
I think I ran into issues when testing `Async.forEachAsync` in the past that this doesn't necessarily end up concurrently executing unless there are multiple async ticks in the loop iteration.

Reply:
For this case, that's fine. I'm not worried about `Async.forEachAsync` itself, but the queue. This test should pass regardless of actually executing concurrently or otherwise. I will add async ticks inline to ensure it goes concurrent.